### PR TITLE
log exceptions caught in Retry

### PR DESCRIPTION
**Pull request by danielkucera**
Log the final exception when either the number of attempts or the timeout were reached.

### CODE REVIEW

**Comment by CyberDem0n** [comment]
Hi @danielkucera,

first of all any PR deserves at least a few words about what the problem is it trying to solve.
And the second, IMHO warning severity is a bit too much for expected exceptions. I am not arguing that having more logs is better than having none, but the severity should be adequate. In this specific case the _debug_ would be enough.

**Comment by danielkucera** [comment]
Hi @CyberDem0n , sorry for missing explanation, I was posting in a hurry.
Retry catches all http exceptions so if you make a mistake in config or e.g. consul server suddenly disappears, you will only see an exception about exhausted retries or timeout. Nothing about the real reason - whethere there is a wrong cert, 503, 404 or whatever.
I think that warning is appropriate because you want to know what happened even if you are running in production with disabled debugging. Exceptions should not normally happen here IMHO.

**Comment by CyberDem0n** [comment]
So far you are the first with such a problem, IMHO debug would be more appropriate.
It is possible to change the log level without restart, and moreover, one can set a different log level [per python module](https://patroni.readthedocs.io/en/latest/SETTINGS.html#log):

```yaml
log:
  level: INFO
  loggers:
    patroni.utils: DEBUG
```

File: patroni/utils.py [hunk 1]
```diff
@@ -308,7 +311,8 @@ def __call__(self, func, *args, **kwargs):
                 if self.deadline is not None and self._cur_stoptime is None:
                     self._cur_stoptime = time.time() + self.deadline
                 return func(*args, **kwargs)
-            except self.retry_exceptions:
+            except self.retry_exceptions as e:
+                logger.warning('Retry got exception: %s', str(e))
```
Comment (CyberDem0n) [inline patroni/utils.py]: `str(e)` us redundant here, just `e` will work absolutely the same.

**Comment by ants** [comment]
I can echo the problem. Having the error output would have made debugging quicker and more approachable in multiple cases. However I was thinking more along the lines of recording the exception and only showing it when retry times out to avoid log spam and, more importantly, warnings about things that are somewhat normal.

**Comment by danielkucera** [comment]
@CyberDem0n , please review the latest version. I tried to make a compromise between our views.

**Comment by CyberDem0n** [comment]
:+1:

**Comment by Jan-M** [comment]
:+1:

**Comment by CyberDem0n** [comment]
Merged, thank you @danielkucera

**Pull request merged**